File: SEMESTER-03/Data Structures/Data Structures/Trees/BinaryTree.cpp
#include <iostream>
#include <string>
#include <queue>

using namespace std;

class node
{
public:
    int data;
    node *left;
    node *right;

    node(int d)
    {
        data = d;
        left = NULL;
        right = NULL;
    }
};
template <class T>
class node2
{
public:
    T data;
    node2 *next;

    node2(T d)
    {
        data = d;
        next = 0;
    }
};

template <class U>
class Queue
{

public:
    node2<U> *front, *rear;

    Queue()
    {
        front = rear = 0;
    }

    void enqueue(U d)
    {

        node2<U> *newnode = new node2<U>(d);

        if (rear == 0)
        {
            front = rear = newnode;
        }
        else
        {

            rear->next = newnode;
            rear = newnode;
        }
    }

    void dequeue()
    {

        if (front == 0)
        {
            cout << "Queue is Empty!";
        }
        else
        {

            node2<U> *temp = front;
            front = front->next;
            free(temp);
        }
    }

    U peek()
    {

        if (!isEmpty())
            return front->data;

        cout << "Queue is Empty!";
        return NULL;
    }

    bool isEmpty()
    {
        if (front == NULL)
        {
            return true;
        }

        return false;
    }
    void display()
    {
        node2<U> *temp = front;
        if (temp == 0)
        {
            cout << "Empty";
        }
        while (temp != 0)
        {
            cout << temp->data << " ";
            temp = temp->next;
        }
    }
};
class BinaryTree
{
public:
    node *root;

    BinaryTree()
    {
        root = 0;
    }

    node *CreateTree()
    {

        int x;

        cout << "Enter data: ";
        cin >> x;

        if (x == -1)
        {
            return NULL;
        }

        node *newnode = new node(x);
        cout << "Enter left of " << x << endl;
        newnode->left = CreateTree();
        cout << "Enter right of " << x << endl;
        newnode->right = CreateTree();

        return newnode;
    }
    void BuildFromLevelOrderTraversal()
    {
        int data;
        cout << "Enter data: ";
        cin >> data;

        root = new node(data);
        queue<node *> q;

        q.push(root);

        while (!q.empty())
        {
            node *temp = q.front();
            q.pop();

            cout << "Enter left of " << temp->data << endl;
            int left;
            cin >> left;
            if (left != -1)
            {
                temp->left = new node(left);
                q.push(temp->left);
            }

            cout << "Enter right of " << temp->data << endl;
            int right;
            cin >> right;
            if (right != -1)
            {
                temp->right = new node(right);
                q.push(temp->right);
            }
        }
    }
    void levelOrderTraversal()
    {

        Queue<node *> q;
        q.enqueue(root);
        q.enqueue(NULL);

        while (!q.isEmpty())
        {
            node *temp = q.peek();
            q.dequeue();

            if (temp == NULL) // using NULL as a seperator taake jab jab NULL aye to hum next line par chale jayenge
            {
                cout << endl;
                if (!q.isEmpty()) // agr queue empty nhi iska matlab ek level or bacha wa ha to uske aage bhi NULL laga denge
                {
                    q.enqueue(NULL);
                }
            }
            else
            {

                cout << temp->data << " ";

                if (temp->left != NULL)
                {
                    q.enqueue(temp->left);
                }

                if (temp->right != NULL)
                {
                    q.enqueue(temp->right);
                }
            }
        }
    }

    void PreOrder(node *root)
    {
        if (root == NULL)
            return;

        cout << root->data << " ";
        PreOrder(root->left);
        PreOrder(root->right);
    }

    void PostOrder(node *root)
    {
        if (root == NULL)
            return;

        PostOrder(root->left);
        PostOrder(root->right);
        cout << root->data << " ";
    }

    void InOrder(node *root)
    {
        if (root == NULL)
            return;

        InOrder(root->left);
        cout << root->data << " ";
        InOrder(root->right);
    }

    node *BuildTreeFromPreorderAndInorder(int lb, int ub, int *inorder, int *preorder, int preorderlen)
    {
        if (lb > ub)
        {
            return NULL;
        }

        if (lb == ub)
        {
            node *newnode = new node(inorder[lb]);
            return newnode;
        }

        int minindex = 100;
        int index = 0;

        for (int i = lb; i <= ub; i++)
        {
            for (int j = 0; j <= preorderlen; j++)
            {
                if (inorder[i] == preorder[j])
                {
                    if (j < minindex)
                    {
                        minindex = j;
                        index = i;
                    }
                    break;
                }
            }
        }
        // cout << preorder[minindex] << endl;
        // cout << "lb: " << lb << " ub: " << ub << endl;
        // cout << "minindex: " << minindex << " index: " << index << endl;
        node *newnode = new node(inorder[index]);

        newnode->left = BuildTreeFromPreorderAndInorder(lb, index - 1, inorder, preorder, preorderlen);
        newnode->right = BuildTreeFromPreorderAndInorder(index + 1, ub, inorder, preorder, preorderlen);

        return newnode;
    }

    node *BuildTreeFromPostorderAndInorder(int lb, int ub, int *inorder, int *postorder, int postorderlen)
    {
        int minindex = -1;
        int index = 0;
        if (lb > ub)
        {
            return NULL;
        }
        if (lb == ub)
        {
            cout << " here" << endl;
            node *newnode = new node(inorder[lb]);
            // cout << inorder[lb] << endl;
            // cout << "lb: " << lb << " ub: " << ub << endl;
            // cout << "minindex: " << minindex << " index: " << index << endl;
            return newnode;
        }

        for (int i = lb; i <= ub; i++)
        {
            for (int j = 0; j <= postorderlen; j++)
            {
                if (inorder[i] == postorder[j])
                {
                    if (j > minindex)
                    {
                        minindex = j;
                        index = i;
                    }
                    break;
                }
            }
        }
        cout << inorder[index] << endl;
        // cout << "lb: " << lb << " ub: " << ub << endl;
        // cout << "minindex: " << minindex << " index: " << index << endl;
        // getchar();
        node *newnode = new node(inorder[index]);

        newnode->left = BuildTreeFromPostorderAndInorder(lb, index - 1, inorder, postorder, postorderlen);
        newnode->right = BuildTreeFromPostorderAndInorder(index + 1, ub, inorder, postorder, postorderlen);

        return newnode;
    }

    // Sir mene is function mei STL waaali QUEUE use ki hai q k merii waali ptaa nhi q nhi chal rhi thi kaafi try kra lekn error smjh nhi aarha tha
    node *Insert(node *root, int val)
    {
        if (root == NULL)
        {
            root = new node(val);
        }
        else
        {
            queue<node *> q;
            q.push(root);

            while (!q.empty())
            {

                node *temp = q.front();
                q.pop();

                if (temp->left == NULL)
                {
                    temp->left = new node(val);
                    break;
                }
                else
                {
                    q.push(temp->left);
                }

                if (temp->right == NULL)
                {
                    temp->right = new node(val);
                    break;
                }
                else
                {
                    q.push(temp->right);
                }
            }
        }

        return root;
    }

    node *Delete(node *root, int val)
    {

        if (root == NULL)
            return NULL;

        if (root->data == val)
        {
            return DeleteNode(root);
        }

        root->left = Delete(root->left, val);
        root->right = Delete(root->right, val);
    }

    node *DeleteNode(node *root)
    {
        if (root->left == NULL && root->right == NULL)
        {
            delete (root);
            return NULL;
        }
        else if (root->left != NULL && root->right == NULL)
        {
            node *temp = root->left;
            delete (root);
            return temp;
        }
        else if (root->left == NULL && root->right != NULL)
        {
            node *temp = root->right;
            delete (root);
            return temp;
        }
        else
        {
            // // this will delete it by value means values will be swapped
            // node *val = max(root->left);
            // int val2 = val->data;
            // root = Delete(root, val->data);
            // root->data = val2;
            // return root;

            // This code is for deleting a node by swapping nodes
            node *prep, *pre;

            prep = root;
            pre = root->left;
            while (pre->right != NULL)
            {
                prep = pre;
                pre = pre->right;
            }

            if (prep != root)
                prep->right = pre->left;
            else
                prep->left = pre->right;

            pre->right = root->right;
            pre->left = root->left;

            delete (root);
            return pre;
        }
    }
    node *search(int val)
    {

        if (root == NULL)
            cout << "Tree is empty!" << endl;

        queue<node *> q;
        q.push(root);

        while (!q.empty())
        {

            node *temp = q.front();
            q.pop();

            if (temp->data == val)
            {
                return temp;
            }

            if (temp->left != NULL)
                q.push(temp->left);

            if (temp->right != NULL)
                q.push(temp->right);
        }

        return NULL;
    }
    node *max(node *root)
    {
        int val1;
        val1 = root->data;
        if (root->right == NULL)
        {
            return root;
        }

        return max(root->right);
    }

    int FindHeight(node *root)
    {

        if (root == NULL)
        {
            return 0;
        }

        int left_height = FindHeight(root->left);
        int right_height = FindHeight(root->right);

        if (left_height > right_height)
            return ++left_height;
        else
            return ++right_height;
    }

    int GetSize(node *root)
    {

        if (root == NULL)
            return 0;

        int left_no_of_nodes = GetSize(root->left);
        int right_no_of_node = GetSize(root->right);

        return (left_no_of_nodes + right_no_of_node) + 1;
    }
};

int main()
{
    BinaryTree b;
    // int inorder[] = {8, 4, 10, 9, 11, 2, 5, 1, 6, 3, 7};
    // int preorder[] = {1, 2, 4, 8, 9, 10, 11, 5, 3, 6, 7};
    // b.root = b.BuildTreeFromPreorderAndInorder(0, 10, inorder, preorder, 10);
    // b.levelOrderTraversal();

    int inorder[] = {9, 5, 1, 7, 2, 12, 8, 4, 3, 11};
    int postorder[] = {9, 1, 2, 12, 7, 5, 3, 11, 4, 8};
    b.root = b.BuildTreeFromPostorderAndInorder(0, 9, inorder, postorder, 9);
    b.levelOrderTraversal();
}

File: SEMESTER-03/Data Structures/Data Structures/Trees/BinaryTree.py
from collections import deque


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class Node2:
    def __init__(self, data):
        self.data = data
        self.next = None


class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

    def enqueue(self, data):
        new_node = Node2(data)

        if self.rear is None:
            self.front = self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node

    def dequeue(self):
        if self.front is None:
            print("Queue is Empty!")
        else:
            temp = self.front
            self.front = self.front.next
            if self.front is None:
                self.rear = None
            del temp

    def peek(self):
        if not self.is_empty():
            return self.front.data

        print("Queue is Empty!")
        return None

    def is_empty(self):
        return self.front is None

    def display(self):
        temp = self.front
        if temp is None:
            print("Empty")
            return

        while temp is not None:
            print(temp.data, end=" ")
            temp = temp.next


class BinaryTree:
    def __init__(self):
        self.root = None

    def create_tree(self):
        x = int(input("Enter data: "))

        if x == -1:
            return None

        new_node = Node(x)
        print(f"Enter left of {x}")
        new_node.left = self.create_tree()
        print(f"Enter right of {x}")
        new_node.right = self.create_tree()

        return new_node

    def build_from_level_order_traversal(self):
        data = int(input("Enter data: "))

        self.root = Node(data)
        q = deque()
        q.append(self.root)

        while q:
            temp = q.popleft()

            print(f"Enter left of {temp.data}")
            left = int(input())
            if left != -1:
                temp.left = Node(left)
                q.append(temp.left)

            print(f"Enter right of {temp.data}")
            right = int(input())
            if right != -1:
                temp.right = Node(right)
                q.append(temp.right)

    def level_order_traversal(self):
        q = Queue()
        q.enqueue(self.root)
        q.enqueue(None)

        while not q.is_empty():
            temp = q.peek()
            q.dequeue()

            if temp is None:  # using None as a separator
                print()
                if not q.is_empty():  # if there's another level
                    q.enqueue(None)
            else:
                print(temp.data, end=" ")

                if temp.left is not None:
                    q.enqueue(temp.left)

                if temp.right is not None:
                    q.enqueue(temp.right)

    def pre_order(self, root):
        if root is None:
            return

        print(root.data, end=" ")
        self.pre_order(root.left)
        self.pre_order(root.right)

    def post_order(self, root):
        if root is None:
            return

        self.post_order(root.left)
        self.post_order(root.right)
        print(root.data, end=" ")

    def in_order(self, root):
        if root is None:
            return

        self.in_order(root.left)
        print(root.data, end=" ")
        self.in_order(root.right)

    def build_tree_from_preorder_and_inorder(
        self, lb, ub, inorder, preorder, preorder_len
    ):
        if lb > ub:
            return None

        if lb == ub:
            new_node = Node(inorder[lb])
            return new_node

        min_index = 100
        index = 0

        for i in range(lb, ub + 1):
            for j in range(preorder_len + 1):
                if inorder[i] == preorder[j]:
                    if j < min_index:
                        min_index = j
                        index = i
                    break

        new_node = Node(inorder[index])

        new_node.left = self.build_tree_from_preorder_and_inorder(
            lb, index - 1, inorder, preorder, preorder_len
        )
        new_node.right = self.build_tree_from_preorder_and_inorder(
            index + 1, ub, inorder, preorder, preorder_len
        )

        return new_node

    def build_tree_from_postorder_and_inorder(
        self, lb, ub, inorder, postorder, postorder_len
    ):
        min_index = -1
        index = 0

        if lb > ub:
            return None

        if lb == ub:
            print(" here")
            new_node = Node(inorder[lb])
            return new_node

        for i in range(lb, ub + 1):
            for j in range(postorder_len + 1):
                if inorder[i] == postorder[j]:
                    if j > min_index:
                        min_index = j
                        index = i
                    break

        print(inorder[index])

        new_node = Node(inorder[index])

        new_node.left = self.build_tree_from_postorder_and_inorder(
            lb, index - 1, inorder, postorder, postorder_len
        )
        new_node.right = self.build_tree_from_postorder_and_inorder(
            index + 1, ub, inorder, postorder, postorder_len
        )

        return new_node

    def insert(self, root, val):
        if root is None:
            root = Node(val)
        else:
            q = deque()
            q.append(root)

            while q:
                temp = q.popleft()

                if temp.left is None:
                    temp.left = Node(val)
                    break
                else:
                    q.append(temp.left)

                if temp.right is None:
                    temp.right = Node(val)
                    break
                else:
                    q.append(temp.right)

        return root

    def delete(self, root, val):
        if root is None:
            return None

        if root.data == val:
            return self.delete_node(root)

        root.left = self.delete(root.left, val)
        root.right = self.delete(root.right, val)

        return root

    def delete_node(self, root):
        if root.left is None and root.right is None:
            del root
            return None
        elif root.left is not None and root.right is None:
            temp = root.left
            del root
            return temp
        elif root.left is None and root.right is not None:
            temp = root.right
            del root
            return temp
        else:
            # This code is for deleting a node by swapping nodes
            prep = root
            pre = root.left

            while pre.right is not None:
                prep = pre
                pre = pre.right

            if prep != root:
                prep.right = pre.left
            else:
                prep.left = pre.right

            pre.right = root.right
            pre.left = root.left

            del root
            return pre

    def search(self, val):
        if self.root is None:
            print("Tree is empty!")
            return None

        q = deque()
        q.append(self.root)

        while q:
            temp = q.popleft()

            if temp.data == val:
                return temp

            if temp.left is not None:
                q.append(temp.left)

            if temp.right is not None:
                q.append(temp.right)

        return None

    def max(self, root):
        if root.right is None:
            return root

        return self.max(root.right)

    def find_height(self, root):
        if root is None:
            return 0

        left_height = self.find_height(root.left)
        right_height = self.find_height(root.right)

        if left_height > right_height:
            return left_height + 1
        else:
            return right_height + 1

    def get_size(self, root):
        if root is None:
            return 0

        left_no_of_nodes = self.get_size(root.left)
        right_no_of_nodes = self.get_size(root.right)

        return left_no_of_nodes + right_no_of_nodes + 1


def main():
    b = BinaryTree()

    # Example 1: Building from preorder and inorder
    # inorder = [8, 4, 10, 9, 11, 2, 5, 1, 6, 3, 7]
    # preorder = [1, 2, 4, 8, 9, 10, 11, 5, 3, 6, 7]
    # b.root = b.build_tree_from_preorder_and_inorder(0, 10, inorder, preorder, 10)
    # b.level_order_traversal()

    # Example 2: Building from postorder and inorder
    inorder = [9, 5, 1, 7, 2, 12, 8, 4, 3, 11]
    postorder = [9, 1, 2, 12, 7, 5, 3, 11, 4, 8]
    b.root = b.build_tree_from_postorder_and_inorder(0, 9, inorder, postorder, 9)
    b.level_order_traversal()


if __name__ == "__main__":
    main()
